File: src/Safir.Agent/Services/FileEventPublisher.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reactive.Linq;
using System.Threading;
using System.Threading.Tasks;
using MediatR;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Safir.Agent.Protos;
using Safir.Messaging;
using Unit = System.Reactive.Unit;

namespace Safir.Agent.Services
{
    internal sealed class FileEventPublisher : IHostedService
    {
        private readonly IFileWatcher _fileWatcher;
        private readonly IPublisher _publisher;
        private readonly ILogger<FileEventPublisher> _logger;
        private List<IDisposable>? _subscriptions;

        public FileEventPublisher(IFileWatcher fileWatcher, IPublisher publisher, ILogger<FileEventPublisher> logger)
        {
            _fileWatcher = fileWatcher ?? throw new ArgumentNullException(nameof(fileWatcher));
            _publisher = publisher ?? throw new ArgumentNullException(nameof(publisher));
            _logger = logger;
        }

        public Task StartAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation("Starting file event publisher");
            _subscriptions = new() {
                _fileWatcher.Created.SelectMany(OnCreated).Subscribe(),
                _fileWatcher.Changed.SelectMany(OnChanged).Subscribe(),
                _fileWatcher.Deleted.SelectMany(OnDeleted).Subscribe(),
                _fileWatcher.Renamed.SelectMany(OnRenamed).Subscribe(),
            };

            return Task.CompletedTask;
        }

        public Task StopAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation("Stopping file event publisher");
            _subscriptions?.ForEach(x => x.Dispose());

            return Task.CompletedTask;
        }

        private Task<Unit> OnCreated(FileSystemEventArgs eventArgs, CancellationToken cancellationToken)
        {
            _logger.LogTrace("Publishing created event");
            var @event = new FileCreated { Path = eventArgs.Name };
            return Publish(_publisher, @event, cancellationToken);
        }

        private Task<Unit> OnChanged(FileSystemEventArgs eventArgs, CancellationToken cancellationToken)
        {
            _logger.LogTrace("Publishing changed event");
            var @event = new FileChanged { Path = eventArgs.Name };
            return Publish(_publisher, @event, cancellationToken);
        }

        private Task<Unit> OnDeleted(FileSystemEventArgs eventArgs, CancellationToken cancellationToken)
        {
            _logger.LogTrace("Publishing deleted event");
            var @event = new FileDeleted { Path = eventArgs.Name };
            return Publish(_publisher, @event, cancellationToken);
        }

        private Task<Unit> OnRenamed(RenamedEventArgs eventArgs, CancellationToken cancellationToken)
        {
            _logger.LogTrace("Publishing renamed event");
            var @event = new FileRenamed { Path = eventArgs.Name, OldPath = eventArgs.OldName };
            return Publish(_publisher, @event, cancellationToken);
        }

        private static async Task<Unit> Publish<T>(
            IPublisher publisher,
            T message,
            CancellationToken cancellationToken)
            where T : IEvent
        {
            await publisher.Publish<T>(message, cancellationToken);
            return Unit.Default;
        }
    }
}

File: src/agent/src/Safir.Agent/Services/FileEventPublisher.fs
namespace Safir.Agent.Services

open System
open System.Threading.Tasks
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging
open Safir.Agent
open Safir.Agent.Services
open System.Reactive.Linq
open Safir.Messaging

type FileEventPublisher(watcher: IFileWatcher, bus: IEventBus, logger: ILogger<FileEventPublisher>) =
    let mutable subscriptions: seq<IDisposable> =
        Seq.empty

    let tryPublish e cancellationToken =
        task {
            try
                logger.LogTrace("Publishing event to bus")
                do! bus.PublishAsync(e, cancellationToken)
            with
            | e -> logger.LogError(e, "Unable to publish event to bus")
        }

    interface IHostedService with
        member this.StartAsync _ =
            do logger.LogTrace("Starting file event publisher")

            subscriptions <-
                [ watcher.Changed
                  |> Observable.map (fun e -> (FileEvent.mapChanged e) :> IEvent)
                  watcher.Created
                  |> Observable.map (fun e -> (FileEvent.mapCreated e) :> IEvent)
                  watcher.Deleted
                  |> Observable.map (fun e -> (FileEvent.mapDeleted e) :> IEvent)
                  watcher.Renamed
                  |> Observable.map (fun e -> (FileEvent.mapRenamed e) :> IEvent) ]
                |> Seq.map (fun o -> o.SelectMany(tryPublish))
                |> Seq.map (fun o -> o.Subscribe())
                |> Seq.toList

            Task.CompletedTask

        member this.StopAsync _ =
            do logger.LogTrace("Stopping file event publisher")
            do subscriptions |> Seq.iter (fun d -> d.Dispose())
            Task.CompletedTask
